### PR TITLE
Populate DependencyTelemetry with Operation Data for Initializers

### PR DESCRIPTION
**Pull request by michaelwstark**
Fix Issue # .
[Consolidated]
https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/900

[Original]
https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/820
https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/747
https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/587

Populate DependencyTelemetry with operation data for initializers.

- [X] I ran Unit Tests locally.

For significant contributions please make sure you have completed the following items:

- [X] Changes in public surface reviewed
- [X] Design discussion issue # https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/587
- [X] CHANGELOG.md updated with one line description of the fix, and a link to the original issue.
- [X] The PR will trigger build, unit tests, and functional tests automatically. If your PR was submitted from fork - mention one of committers to initiate the build for you.
	  If you want to to re-run the build/tests, the easiest way is to simply Close and Re-Open this same PR. (Just click 'close pull request' followed by 'open pull request' buttons at the bottom of the PR)

- Please follow [these] (https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/CONTRIBUTING.md) instructions to build and test locally.

### CODE REVIEW

**Comment by michaelwstark** [comment]
I assume there is a change to the core SDK nuget package required, but as that hasn't been built yet, I just worked with the update locally.  If there is a better processor for changes which modify both, please point me in the right direction -- appreciated.

File: Src/DependencyCollector/Shared/HttpCoreDiagnosticSourceListener.cs [hunk 1]
```diff
@@ -350,6 +351,8 @@ internal void OnActivityStop(HttpResponseMessage response, HttpRequestMessage re
             if (response != null)
             {
                 this.ParseResponse(response, telemetry);
+                telemetry.OperationDetails[RemoteDependencyConstants.HttpResponseOperationDetailName] = response;
+                telemetry.OperationDetails[RemoteDependencyConstants.HttpResponseHeadersOperationDetailName] = response.Headers;
```
Comment (lmolkova) [inline Src/DependencyCollector/Shared/HttpCoreDiagnosticSourceListener.cs]: why do we need to populate response AND headers? I'd prefer them to be mutually exclusive. 
In case of .NET Core there is no case when a response is missing, but headers are present

File: Src/DependencyCollector/Shared/Implementation/SqlClientDiagnostics/SqlClientDiagnosticSourceListener.cs [hunk 2]
```diff
@@ -116,7 +118,14 @@ public void Dispose()
                                 Data = command.CommandText,
                                 Success = true
                             };
-                            
+
+                            // Populate the operation details for intializers
+                            telemetry.OperationDetails[RemoteDependencyConstants.SqlCommandOperationDetailName] = command;
+                            if (connection != null)
+                            {
+                                telemetry.OperationDetails[RemoteDependencyConstants.SqlConnectionOperationDetailName] = connection;
```
Comment (lmolkova) [inline Src/DependencyCollector/Shared/Implementation/SqlClientDiagnostics/SqlClientDiagnosticSourceListener.cs]: The connection is available through the command, why do we need to explicitly add it to the dictionary?

**Comment by lmolkova** [comment]
@anpete 
Could you please help us review these SQL dependency collection changes?

Our goal is to enable scenario when users add arbitrary dynamic properties to their dependency telemetry. In HTTP case it could be adding request or response header value to the telemetry.

Could you please help us understand whether it is relevant for SQL and what kind of information could be useful/not useful.

Basically, we proxy DiagnostiSource payload to users through telemetry instance (but this info is not serialized and is never sent to the backend as it). It's up to users to access these objects and modify telemetry based on them.

This is a very advanced scenario we don't expect the absolute majority of users to implement.

File: Src/DependencyCollector/Shared/HttpCoreDiagnosticSourceListener.cs (hunk 1 above)
Comment (michaelwstark) [reply Src/DependencyCollector/Shared/HttpCoreDiagnosticSourceListener.cs]: That was actually how I originally had it (until I hit the corner case where the response was not present).  My rationale for always including them was to make development of a initializer easier -- if the developer is only interested in headers, headers is always populated.  This is not a stance of principal or conviction -- you are much more familiar with what the developer experience of the initializers should be, so I will defer to you.  Please ack that you would prefer to only have the headers when the response is not available and I will make the change after I finish an update on the SDK pull request.

File: Src/DependencyCollector/Shared/Implementation/SqlClientDiagnostics/SqlClientDiagnosticSourceListener.cs (hunk 2 above)
Comment (michaelwstark) [reply Src/DependencyCollector/Shared/Implementation/SqlClientDiagnostics/SqlClientDiagnosticSourceListener.cs]: Similar rationale as the response\headers comment above.  I was not sure that the command would always be populated based on the failure cases, and I wanted to make it easier to develop the initializer if someone was only interested in properties on the connection.  I will wait for confirmation on the comment above, and align the two to be consistent.

File: Src/DependencyCollector/Shared/Implementation/SqlClientDiagnostics/SqlClientDiagnosticSourceListener.cs [hunk 3]
```diff
@@ -301,7 +306,7 @@ public void Dispose()
                                 Type = RemoteDependencyConstants.SQL,
                                 Target = string.Join(" | ", connection.DataSource, connection.Database),
                                 Data = operation,
-                                Success = true
+                                Success = true,
```
Comment (cijothomas) [inline Src/DependencyCollector/Shared/Implementation/SqlClientDiagnostics/SqlClientDiagnosticSourceListener.cs]: intentional?

File: Src/DependencyCollector/Shared/Implementation/SqlClientDiagnostics/SqlClientDiagnosticSourceListener.cs [hunk 4]
```diff
@@ -116,7 +118,10 @@ public void Dispose()
                                 Data = command.CommandText,
                                 Success = true
                             };
-                            
+
+                            // Populate the operation details for intializers
+                            telemetry.SetOperationDetail(RemoteDependencyConstants.SqlCommandOperationDetailName, command);
```
Comment (cijothomas) [inline Src/DependencyCollector/Shared/Implementation/SqlClientDiagnostics/SqlClientDiagnosticSourceListener.cs]: need similar change here as well:
https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/Src/DependencyCollector/Shared/Implementation/ProfilerSqlProcessingBase.cs
This contains SQL dependency collection logic with Profiler installed.

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 ## Version 2.7.0-beta1
 - [Added support to collect Perf Counters for .NET Core Apps if running inside Azure WebApps] (https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/889)
+- [Add operation details for HTTP and SQL operation to the dependency telemetry.](https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/587)
```
Comment (cijothomas) [inline CHANGELOG.md]: If possible create a new issue in gh, which contain detail about the specific fix. The original issue is quite long discussion, and it'll be hard to figure out which part from that is being implemented here.

Comment (cijothomas) [reply CHANGELOG.md]: Also helpful if an example TelemetryInitializer can be provided which utilizes this new feature.

**Comment by cijothomas** [comment]
@SergeyKanzhelev  and @Dmitry-Matveev - Please share your thoughts on making this under a feature flag, and having them OFF by default. I'd assume that this is only useful if customer has their own telemetry initializer which makes use of this. Those customers can turn this feature ON, so regular users dont see increased memory usage from AI SDK.

**Comment by michaelwstark** [comment]
@cijothomas - While we are waiting for @SergeyKanzhelev and @Dmitry-Matveev to respond to whether this needs to be placed under a feature flag, can you point me to a current example?  I presume it would be something similar to ExcludeComponentCorrelationHttpHeadersOnDomains or IncludeDiagnosticSourceActivities on the current dependency collection module, but am not completely sure.

File: Src/DependencyCollector/Shared/Implementation/SqlClientDiagnostics/SqlClientDiagnosticSourceListener.cs (hunk 3 above)
Comment (michaelwstark) [reply Src/DependencyCollector/Shared/Implementation/SqlClientDiagnostics/SqlClientDiagnosticSourceListener.cs]: Not intentional, but it is a coding style from my projects.  I'll remove the trailing comma when we hear back regarding the feature flag.

File: CHANGELOG.md (hunk 5 above)
Comment (cijothomas) [reply CHANGELOG.md]: Either is fine. Inline in a GH issue is easier to find, in my opinion though.

**Comment by cijothomas** [comment]
@michaelwstark Yes you are pretty close. More relevant example is `DisableDiagnosticSourceInstrumentation` flag on `DependencyTrackingTelemetryModule` which turns on/off DiagnosticSource based dependency tracking. 
https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/Src/DependencyCollector/Shared/DependencyTrackingTelemetryModule.cs#L57

File: CHANGELOG.md (hunk 5 above)
Comment (michaelwstark) [reply CHANGELOG.md]: K.  I'll code up a sample as I was intending to use this feature and include it in a new GH issue.

**Comment by michaelwstark** [comment]
Thanks for the pointer.  I will create a branch off of my current working code with this change in anticipation of @SergeyKanzhelev or @Dmitry-Matveev indicating that we should include a flag for this behavior.

**Comment by lmolkova** [comment]
> Those customers can turn this feature ON, so regular users dont see increased memory usage from AI SDK.

The should be no increase in memory usage. The properties DependencyTelemetry obtains already present in the memory and are cleaned up after dependency call is finished.

E.g. in case of HTTP we add Http request and Http response - there are no new allocations, except for dictionary with 2 items.

**Comment by lmolkova** [comment]
@michaelwstark 

we discussed it with @cijothomas offline.
The concern is that the feature would postpone  GC for the requests/responses since reference would be kept in dependency telemetry until it is sent to the backend (that happens asynchronously).

I wonder, can we clean up the OperationDetails in `TrackDependency` call after initialize is called to address this concern?

[Here](https://github.com/Microsoft/ApplicationInsights-dotnet/blob/9479e37efc5221f3b8e238696be1050128d21b2c/src/Microsoft.ApplicationInsights/TelemetryClient.cs#L370)
```csharp
        public void TrackDependency(DependencyTelemetry telemetry)
        {
            if (telemetry == null)
            {
                telemetry = new DependencyTelemetry();
            }

            this.Track(telemetry);

!!!         telemetry.OperationDetails.Clear();
        }
```

It totally makes sense to me, as Track happens syncronously with dependency call and request/response should not be accessed after that.

File: CHANGELOG.md (hunk 5 above)
Comment (michaelwstark) [reply CHANGELOG.md]: Here is a consolidated issue with a code sample for intended use:
https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/900

**Comment by Dmitry-Matveev** [comment]
@michaelwstark, I'm with Cijo on this one - due to increased memory usage for default scenario, I'd suggest to make this feature opt-in rather than on by default. Performant apps (let's say 500+ RPS) may all of a sudden become less performant if we store all request/responses/headers for 30 seconds (15000 * 3 objects). Super-performant apps (1000+ RPS) may struggle even if we come up with some optimization to release object sooner - it will still exist more than ASP.NET intended it to.
Making this opt-in will require a switch or flag that you are discussing with Cijo already.

**Comment by lmolkova** [comment]
Chatted with @Dmitry-Matveev offline (he had to run for a meeting so cannot reply right now).
He is fine with the approach when we clean up details immediately after tracking telemetry (sync with response events). He agrees that in this case feature flag is not needed.

**Comment by Dmitry-Matveev** [comment]
Thanks, @lmolkova! @michaelwstark, in case we do not copy nor keep response/request objects after the request is over - we are good without the feature flag.

**Comment by michaelwstark** [comment]
@Dmitry-Matveev Thanks!  @lmolkova - I think that all of the outstanding feedback has been resolved, please let me know what else needs to be done prior to merging these PRs into their respective repos.  I noticed there are some VSTS checks which never provide any feedback -- anything I need to do there?

**Comment by cijothomas** [comment]
@michaelwstark thanks for addressing the suggestions!
The VSTS checks are for CI Builds, which are not auto-triggered for forked repo pull requests. I'll manually trigger them for you and if all tests pass, we are good.
The changes in base SDK need to be first merged and published before the changes in this repo can be merged. I will start in base sdk build trigger first.

**Comment by cijothomas** [comment]
@MS-TimothyMothra to ensure this is merged **AFTER** base 2.7.0-beta1 is published to myget.

**Comment by MS-TimothyMothra** [comment]
Sorry, I misunderstood and prep-ed this for Beta2.
Reverted my change.

**Comment by michaelwstark** [comment]
@MS-TimothyMothra has the Beta1 of the core SDK been published to myget?  When do we expect this PR to merge?  Thanks in advance.

**Comment by MS-TimothyMothra** [comment]
Sorry @michaelwstark, the 2.7 Beta1 got delayed to June 1st because we're working on hotfixes for 2.6 Stable.
Feel free to contact me directly if you need more information

**Comment by MS-TimothyMothra** [comment]
@michaelwstark It looks like some of your new tests are failing.

- TestZeroAndNonZeroContentResponseDiagnosticSource
- TestBasicDependencyCollectionDiagnosticSourceLegacyHeaders

Will you have time to look at these today?

**Comment by michaelwstark** [comment]
@MS-TimothyMothra.  Yup.  I'll be able to take a look today.

**Comment by lmolkova** [comment]
I've added these two recently:

TestZeroAndNonZeroContentResponseDiagnosticSource
TestBasicDependencyCollectionDiagnosticSourceLegacyHeaders

Probably failures are result of auto-merge and new tests are missing something.

**Comment by michaelwstark** [comment]
One test is a negative case for response, and the validate call requires hint not to validate response data.  The other leveraged a new Boolean value in the validate pipeline which was not named which conflicted with the one I added.  Re-running unit tests after the update and will push shortly if they pass.

**Comment by michaelwstark** [comment]
@MS-TimothyMothra you should be good to proceed now.  Let me know if you hit any more issues.

**Comment by MS-TimothyMothra** [comment]
Thanks @michaelwstark ! Queuing your build now.
